saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/registration/OpenSamlAssertingPartyMetadataConverter.java

```diff
@@ -77,44 +67,28 @@
  * @author Josh Cummings
  * @since 5.4
  */
-public class OpenSamlRelyingPartyRegistrationBuilderHttpMessageConverter
-		implements HttpMessageConverter<RelyingPartyRegistration.Builder> {
+public class OpenSamlAssertingPartyMetadataConverter {
```

Comment:
I think this should be two classes. One is something that implements `HttpMessageConverter` and one that just deals with the logic of converting an `InputStream` into a `RelyingPartyRegistration.Builder`. This existing class's public contract would remain as-is and the new class would be called `OpenSamlAssertingPartyMetadataConverter`. This existing class would call the new class.

Reply:
But won't the use of the `ResourceLoader` always return an `InputStream` instead of an `HttpInputMessage` with this new approach? Irrespective if it's a file or a url location that is what i have seen from the tests i have done. This was the reason i removed the `HttpMessageConverter` completely.

Reply:
The `HttpMessageConverter` class is still valuable for applications that are using `RestTemplate` to retrieve the resource. 

While the basic `UrlResourceLoader` may work in the most basic HTTP cases, any sophisticated tweaking -- connection timeouts, proxy configuration, etc. -- will require using `RestTemplate`.

Reply:
Sure, no problem.

There are two use cases that these classes cater to:

* Applications that want to supply a URI and get a `RelyingPartyRegistration`. That's fulfilled by `RelyingPartyRegistrations`.
* Applications that need a `RestTemplate` to get a `RelyingPartyRegistration`. That's fulfilled by `OpenSamlRelyingPartyRegistrationBuilderHttpMessageConverter`, in conjunction with a `RestTemplate` that the application itself would instantiate.

Both of these classes - `RelyingPartyRegistrations` and `OpenSamlRelyingPartyRegistrationBuilderHttpMessageConverter` - are going to have an `InputStream` that they want converted into a `RelyingPartyRegistration.Builder`, so they will each have a private instance of `OpenSamlAssertingPartyMetadataConverter`.

When this change is complete, `RelyingPartyRegistrations` will no longer use a `RestTemplate`, which means it will no longer call `OpenSamlRelyingPartyRegistrationBuilderHttpMessageConverter`.

Does that clarify?

Reply:
Understood and now it makes total sense. I have restored the `OpenSamlRelyingPartyRegistrationBuilderHttpMessageConverter` and the parsing logic is now extracted into `OpenSamlAssertingPartyMetadataConverter`. I left the tests for the Http converter and added a class that also tests for the Url or File based metadata files.